keps/sig-apps/3715-indexed-job-with-unset-completions/README.md

```diff
@@ -0,0 +1,913 @@
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+<!--
+**ACTION REQUIRED:** In order to merge code into a release, there must be an
+issue in [kubernetes/enhancements] referencing this KEP and targeting a release
+milestone **before the [Enhancement Freeze](https://git.k8s.io/sig-release/releases)
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core
+Kubernetes—i.e., [kubernetes/kubernetes], we require the following Release
+Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [x] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [x] (R) Design details are appropriately documented
+- [x] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [x] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Currently Jobs are validated to have `spec.completions` set when `spec.completionMode`=`Indexed`. This KEP proposes 
+to relax this validation and allow Indexed jobs to be created with unset `spec.completions`. 
+
+When `spec.completions` is not set for an `Indexed` Job, the proposed behavior is summarized as follows:
+- The success, failure and finish semantics are same as `NonIndexed`. Specifically:
```

Comment:
@tenzen-y can you confirm if the existing success semantics will work for elastic Pytorch and Horovod? 

My understanding is that the "manager" decides the success of the workload as a whole, and we want to model the manager as a separate job from the workers. So the exact success policy doesn't matter much.

Reply:
Thanks for mentioning me.

> My understanding is that the "manager" decides the success of the workload as a whole, and we want to model the manager as a separate job from the workers.

Yes, that's right for Horovod. In kubeflow Elastic PytorchJob, we launch the "manager" (in PyTorch, "manager" is called the elastic agent) on the Container, the same as the training Container as a subprocess, not a sidecar container or another Pod.

Although, I guess that existing success semantics would not be a bad influence on kubeflow Elastic PytorchJob since all Workers have an elastic agent.